etna/records/models.py

```diff
@@ -1,78 +1,316 @@
+
+    @cached_property
+    def iaid(self) -> str:
+        """
+        Return the "iaid" value for this record (if one is available).
+
+        Raises `ValueExtractionError` when the raw data does not include
+        any candidate values.
+
+        Raises `ValueError` when the raw data includes a value where iaids
+        are usually found, but the value is not a valid iaid.
+        """
+        try:
+            candidate = self.template["iaid"]
+        except KeyError:
+            candidate = self.get("@admin.id")
+
+        # value is not guaranteed to be a valid 'iaid', so we must
+        # check it before returning it as one
+        if not re.match(IAIDConverter.regex, candidate):
+            raise ValueError(f"Value '{candidate}' from API is not a valid iaid.")
+        return candidate
+
+    def has_iaid(self) -> bool:
+        """
+        Returns `True` if a valid 'iaid' value can be extracted from the
+        raw data for this record. Otherwise `False`.
+        """
+        try:
+            self.iaid
+        except (ValueExtractionError, ValueError):
+            return False
+        else:
+            return True
+
+    @cached_property
+    def reference_number(self) -> str:
+        """
+        Return the "reference_number" value for this record (if one is available).
+
+        Raises `ValueExtractionError` when the raw data does not include
+        values in any of the expected positions.
+        """
+        try:
+            return self.template["referenceNumber"]
+        except KeyError:
+            pass
+        identifiers = self.get("identifier", ())
+        for item in identifiers:
+            try:
+                return item["reference_number"]
+            except KeyError:
+                pass
+        raise ValueExtractionError(
+            f"'reference_number' could not be extracted from source data: {self._raw}"
+        )
+
+    def has_reference_number(self) -> bool:
+        """
+        Returns `True` if a 'reference_number' value can be extracted from the
+        raw data for this record. Otherwise `False`.
+        """
+        try:
+            self.reference_number
+        except ValueExtractionError:
+            return False
+        else:
+            return True
+
+    @cached_property
+    def url(self):
+        """
+        Return the "url" value for this record. This value is typically
+        only present for 'interpretive' results from other websites.
+
+        Raises `ValueExtractionError` when the raw data does not include
+        values in any of the expected positions.
+        """
+        try:
+            return self.template["sourceUrl"]
+        except KeyError:
+            pass
+
+        locations = self.get("source.@location", ())
+        if len(locations) == 1:
+            return locations[0]["value"]
+        for item in locations:
+            if item.get("public", False):
+                return item["value"]
```

Comment:
Could not find this path `source.@location` in api response or spec using stream=interpretive. 
Can we have a test case for this.

Reply:
Hi @TNA-Allan. These are present for insights results:
{{BASE_URL}}/api/v1/data/searchAll?filterAggregations=group:insight&size=10

Reply:
Have removed in favour of just looking for `self.template["sourceUrl"]`